File: Python.js
/*
 * Python 3 Regular Expressions
 *
 * @author Peter Varo
 * @version 2.0.0
 */

var self = [
//- COMMENT ------------------------------------------------------------------//
    {
        'name' : 'comment.line.hashmark.python',
        'pattern': /#.*/g
    },


//- LAMBDA -------------------------------------------------------------------//
    {
        'pattern': /\b(lambda)(\s+(\*{0,2}[A-z_]\w*\s*(,|=.+(?=,|:))?\s*)*)?(?=:)/g,
        'matches':
        {
            1: 'storage.type.function.anonymous.python',
            2:
            {
                'pattern': /\*{0,2}([A-z_]\w*)\s*(,|=.+(?=,|:))?\s*/g,
                'matches':
                {
                    1: 'variable.parameter.function.python',
                    2:
                    {
                        'pattern': /(=)(.+)(?=,|:)/g,
                        'matches':
                        {
                            1: 'keyword.operator.assignment.python',
                            2:
                            {
                                'language': 'python'
                            }
                        }
                    }
                }
            }
        }
    },


//- NUMBERS ------------------------------------------------------------------//
    {
        'name' : 'constant.numeric.integer.binary.python',
        'pattern': /\b0b[01]+/g
    },
    {
        'name' : 'constant.numeric.integer.hexadecimal.python',
        'pattern': /\b0x[0-9a-fA-F]+/g
    },
    {
        'name' : 'constant.numeric.integer.octal.python',
        'pattern': /\b0o[0-7]+/g
    },
    {
        //.001  .1e6  .1E6  .1e+6  .1E+6  .1e-6  .1E-6
        'name' : 'constant.numeric.float_and_complex.decimal.floatnumber.python',
        'pattern': /(?=\W)\.\d+([eE][+-]?\d+)?[jJ]?/g
    },
    {
        //1.  1.0  1.1e6  1.1E6  1.1e+6  1.1E+6  1.1e-6  1.1E-6
        'name' : 'constant.numeric.float_and_complex.decimal.pointfloat.python',
        'pattern': /\d+\.(\d+([eE][+-]?\d+)?)?[jJ]?/g
    },
    {
        //1e6  1E6  1e+6  1E+6  1e-6  1E-6
        'name' : 'constant.numeric.float_and_complex.decimal.exponent.python',
        'pattern': /\b(?!\.)\d+[eE][+-]?\d+[jJ]?/g
    },
    {
        'name' : 'constant.numeric.integer_and_complex.decimal.python',
        'pattern': /\b(?!\.)([1-9]\d*|0)[jJ]?/g
    },


//- KEYWORDS -----------------------------------------------------------------//
    {
        'name' : 'storage.modifier.declaration.python',
        'pattern': /\b(global|nonlocal)\b/g
    },
    {
        'name' : 'keyword.control.import_and_import_from.python',
        'pattern': /\b(import|from)\b/g
    },
    {
        'name' : 'keyword.control.flow_block_delimiters.python',
        'pattern': /\b(elif|else|except|finally|for|if|try|while|with|break|continue|pass|raise|return|yield)\b/g
    },
    {
        'name' : 'keyword.operator.bool.logical.python',
        'pattern': /\b(and|in|is|not|or)\b/g
    },
    {
        'name' : 'keyword.other.python',
        'pattern': /\b(as|assert|del)\b/g
    },


//- OPERATORS ----------------------------------------------------------------//
    {
        'name' : 'keyword.operator.comparison.python',
        'pattern': /&lt;=|&gt;=|==|&lt;|&gt;|!=/g
    },
    {
        'name' : 'keyword.operator.assignment.augmented.python',
        'pattern': /\+=|-=|\*=|\/=|\/\/=|%=|&amp;=|\|=|\^=|&lt;&lt;=|&gt;&gt;=|\*\*=/g
    },
    {
        'name' : 'keyword.operator.arithmetic.python',
        'pattern': /\+|-|\*|\*\*|\/|\/\/|%|&lt;&lt;|&gt;&gt;|&amp;|\||\^|~/g
    },
    {
        'name' : 'keyword.operator.value_and_annotation_assignment.python',
        'pattern': /=|-&gt;/g
    },


//- CLASS --------------------------------------------------------------------//
    {
        'pattern': /\s*(class)\s+([A-z_]\w*)(\s*\((([A-z_]\w*\s*,?\s*)|([A-z_]\w*\s*=\s*[A-z_]\w*))*\))?:/g,
        'matches':
        {
            1: 'storage.type.class.python',
            2: 'entity.name.type.class.python',
            3:
            {
                'pattern': /([A-z_]\w*\s*,?\s*)|([A-z_]\w*\s*=\s*[A-z_]\w*)/g,
                'matches':
                {
                    1:
                    {
                        'pattern': /([A-z_]\w*)\s*,?\s*/g,
                        'matches':
                        {
                            //todo: invalid.illegal names (reserved keywords)
                            1: 'entity.other.inherited-class.python'
                        }
                    },
                    2:
                    {
                        'pattern': /([A-z_]\w*)\s*(=)\s*([A-z_]\w*)/g,
                        'matches':
                        {
                            1: 'entity.other.inherited-class.python',
                            2: 'keyword.operator.assignment.python',
                            3: 'entity.other.inherited-class.python'
                        }
                    }
                }
            }
        }
    },


//- FUNCTION -----------------------------------------------------------------//
    {
        //'pattern': /\s*(def)\s+([A-z_]\w*)\s*\((([A-z_]\w*\s*(?=,|[\n)])\s*)*(\*[A-z_]\w*\s*(?=,|[\n)])\s*)?([A-z_]\w*\s*=\s*.+(?=,|[\n)])\s*)*(\*\*[A-z_]\w*(?=[\n)])\s*)?)\):/g,
        'pattern': /\s*(def)\s+([A-z_]\w*)\s*\((\*?\*?([A-z_]\w*)\s*(:.+(?=,|=|[\n)]))?\s*(=.+(?=,|[\n)]))?\s*)*\)\s*(-&gt;.+)?(?=:)/g,
        'matches':
        {
            1: 'storage.type.function.python',
            2: 'entity.name.function.python',
            3:
            {
                'pattern': /\*?\*?([A-z_]\w*)\s*(?=:.+(?=,|=|[\n)]))?\s*(?==.+(?=,|[\n)]))?\s*/g,
                'matches':
                {
                    1: 'variable.parameter.function.python'
                }
            },
            4:
            {
                'pattern': /(-&gt;).+/g,
                'matches':
                {
                    1: 'keyword.operator.assignment.python'
                }
            }
            // 3:
            // {
            //     'pattern': /([A-z_]\w*\s*(?=,|[\n)])\s*)*(\*[A-z_]\w*\s*(?=,|[\n)])\s*)?([A-z_]\w*\s*=\s*.+(?=,|[\n)])\s*)*(\*\*[A-z_]\w*(?=[\n)])\s*)?/g,
            //     'matches':
            //     {
            //         1:
            //         {
            //             'pattern': /([A-z_]\w*)\s*(?=,|[\n)])\s*/g,
            //             'matches':
            //             {
            //                 1: 'variable.parameter.function.python',
            //                 3: /(\*{0,2}([A-z_]\w*)\s*(:.+(?=,|=|[\n)]))?\s*(=.+(?=,|[\n)]))?)*/
            //             }
            //         },
            //         2:
            //         {
            //             'pattern': /\*([A-z_]\w*)\s*(?=,|[\n)])\s*/g,
            //             'matches':
            //             {
            //                 1: 'variable.parameter.function.python'
            //             }
            //         },
            //         3:
            //         {
            //             'pattern': /([A-z_]\w*)\s*=\s*.+(?=,|[\n)])\s*/g,
            //             'matches':
            //             {
            //                 1: 'variable.parameter.function.python',
            //                 2: 'keyword.operator.assignment.python'
            //             }
            //         },
            //         4:
            //         {
            //             'pattern': /\*\*([A-z_]\w*)(?=[\n)])\s*/g,
            //             'matches':
            //             {
            //                 1: 'variable.parameter.function.python'
            //             }
            //         }
            //     }
            // }
        }
    }
];

Rainbow.extend('python', self, true);

File: src/python.py
## INFO ########################################################################
##                                                                            ##
##                   Python and Cython Syntax Highlighters                    ##
##                   =====================================                    ##
##                                                                            ##
##                       Version: 2.0.00.015 (20141006)                       ##
##                            File: src/python.py                             ##
##                                                                            ##
##            For more information about the project, please visit            ##
##                   <https://github.com/petervaro/python>.                   ##
##                    Copyright (C) 2013 - 2014 Peter Varo                    ##
##                                                                            ##
##  This program is free software: you can redistribute it and/or modify it   ##
##   under the terms of the GNU General Public License as published by the    ##
##       Free Software Foundation, either version 3 of the License, or        ##
##                    (at your option) any later version.                     ##
##                                                                            ##
##    This program is distributed in the hope that it will be useful, but     ##
##         WITHOUT ANY WARRANTY; without even the implied warranty of         ##
##            MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.            ##
##            See the GNU General Public License for more details.            ##
##                                                                            ##
##     You should have received a copy of the GNU General Public License      ##
##     along with this program, most likely a file in the root directory,     ##
##        called 'LICENSE'. If not, see <http://www.gnu.org/licenses>.        ##
##                                                                            ##
######################################################################## INFO ##

#-- CHEATSHEET ----------------------------------------------------------------#
# HOWTO: http://sublimetext.info/docs/en/reference/syntaxdefs.html
# REGEX: http://manual.macromates.com/en/regular_expressions

# Import python modules
from copy import deepcopy

# Import user modules
from src.common import syntax as original

# Syntax Definition
syntax = deepcopy(original)

# Auto-recognition
syntax['fileTypes'] = ['py']
syntax['keyEquivalent'] = '^~P'

# Shebang
syntax['firstLineMatch'] = r'^#!/.*\bpython[\d.-]*\b'

# Folding marks for the TextEditor
syntax['foldingStartMarker'] = (r'^\s*(def|class)\s+([.\w>]+)\s*(\((.*)\))?\s*:'
                                r'|\{\s*$|\(\s*$|\[\s*$|^\s*"""(?=.)(?!.*""")')
syntax['foldingStopMarker'] = r'^\s*$|^\s*\}|^\s*\]|^\s*\)|^\s*"""\s*$'

# Language ID
syntax['uuid'] = '851B1429-B8B4-4C1E-8030-399BDA994393'

# Patterns
syntax['patterns'].update({
#-- COMMENT -------------------------------------------------------------------#
        # ALL ARE COMMON

#-- KEYWORDS ------------------------------------------------------------------#
        0x0080:
        {
            'name' : 'storage.modifier.declaration.{SCOPE}',
            'match': r'\b(global|nonlocal)\b'
        },

        0x0090:
        {
            'name' : 'keyword.control.import_and_import_from.{SCOPE}',
            'match': r'\b(import|from)\b'
        },

        0x00A0:
        {
            'name' : 'keyword.control.flow_block_delimiters.{SCOPE}',
            'match':
            (
                r'\b(elif|else|except|finally|for|if|try|while|'
                r'with|break|continue|pass|raise|return|yield)\b'
            )
        },

        0x00C0:
        {
            'name' : 'keyword.other.{SCOPE}',
            'match': r'\b(as|assert|del)\b'
        },


#-- OPERATORS -----------------------------------------------------------------#
        # ALL ARE COMMON

#-- CLASS ---------------------------------------------------------------------#
        0x0110:
        {
            'name' : 'meta.class.{SCOPE}',
            'begin': r'^\s*(class)\s+(?=[a-zA-Z_]\w*(\s*\()?)',
            'beginCaptures':
            {
                1: {'name': 'storage.type.class.{SCOPE}'}
            },
            'patterns':
            [
                {'include': '#class_entity_name'},
                {'include': '#class_inheritance'}
            ],
            'end'  : r'(\)?\s*:|\s+([\w#\s:]+))',
            'endCaptures':
            {
                3: {'name': 'invalid.illegal.missing_section_begin.{SCOPE}'}
            }
        },


#-- FUNCTION ------------------------------------------------------------------#
        0x0120:
        {
            'name' : 'meta.function.{SCOPE}',
            'begin': r'^\s*(def)\s+(?=[a-zA-Z_]\w*\s*\()',
            'beginCaptures':
            {
                1: {'name': 'storage.type.function.{SCOPE}'}
            },
            'patterns':
            [
                # Function name
                {'include': '#function_entity_name'},
                # Arguments
                {'include': '#function_arguments'},
                # Annotation assignment (function)
                {'include': '#function_annotation'}
            ],
            # TODO: add illegal
            'end': r'(\s*:)',
            'endCaptures':
            {
                2: {'name': 'invalid.illegal.missing_section_begin.{SCOPE}'}
            }
        },


#-- LAMBDA --------------------------------------------------------------------#
        # ALL ARE COMMON

#-- DECORATOR -----------------------------------------------------------------#
        # ALL ARE COMMONE

#-- CONSTANTS -----------------------------------------------------------------#
        0x0160:
        {
            'name' : 'constant.language.word_like.{SCOPE}',
            'match': r'\b(None|True|False|Ellipsis|NotImplemented)\b'
        },


#-- STORAGES ------------------------------------------------------------------#
        0x0180:
        {
            'name' : 'storage.type.function.{SCOPE}',
            'match': r'\b(def|lambda)\b'
        },


#-- BUILTINS ------------------------------------------------------------------#
        # ALL ARE COMMON

#-- MAGIC STUFFS --------------------------------------------------------------#
        # ALL ARE COMMON

#-- ETC -----------------------------------------------------------------------#
        # ALL ARE COMMON

#-- STRUCTURES ----------------------------------------------------------------#
        # ALL ARE COMMON

#-- ACCESS --------------------------------------------------------------------#
        # ALL ARE COMMON

#-- STRING --------------------------------------------------------------------#
        # ALL ARE COMMON
    })


#-- REPOSITORY ----------------------------------------------------------------#
syntax['repository'].update({
#-- COMMENTS ------------------------------------------------------------------#
        # ALL ARE COMMON

#-- CLASS ---------------------------------------------------------------------#
        # ALL ARE COMMON

#-- FUNCTION ------------------------------------------------------------------#
        # ALL ARE COMMON

#-- BUILTINS ------------------------------------------------------------------#
        # TODO: rearrange -> what is builtin function and what is builtin type?
        'builtin_functions':
        {
            'name' : 'support.function.builtin.{SCOPE}',
            'match':
            (
                r'(?<!\.)\b('
                r'__import__|abs|all|any|ascii|bin|callable|chr|compile|delattr|'
                r'dir|divmod|eval|exec|filter|format|getattr|globals|hasattr|hash|'
                r'help|hex|id|input|isinstance|issubclass|iter|len|locals|map|max|'
                r'min|next|oct|ord|pow|print|range|repr|round|setattr|sorted|sum|'
                r'vars|zip'
                r')\b'
            )
        },
        'builtin_types':
        {
            'name' : 'support.type.{SCOPE}',
            'match':
            (
                r'(?<!\.)\b('
                r'basestring|bool|bytearray|bytes|classmethod|complex|dict|'
                r'enumerate|float|frozenset|int|list|memoryview|object|open|'
                r'property|reversed|set|slice|staticmethod|str|super|tuple|type'
                r')\b'
            )
        },


#-- ENTITY --------------------------------------------------------------------#
        'illegal_names':
        {
            'name' : 'invalid.illegal_names.name.{SCOPE}',
            'match':
            (
                r'\b('
                r'and|as|assert|break|class|continue|def|del|elif|else|except|'
                r'finally|for|from|global|if|import|in|is|lambda|nonlocal|not|'
                r'or|pass|raise|return|try|while|with|yield'
                r')\b'
            )
        },


#-- KEYWORDS ------------------------------------------------------------------#
        # ALL ARE COMMON

#-- MAGIC STUFFS --------------------------------------------------------------#
        # TODO: rearrange -> what is magic function and what is magic variable?
        'magic_function_names':
        {
            'name' : 'support.function.magic.{SCOPE}',
            'match':
            (
                r'\b__('
                r'abs|add|and|bool|bytes|call|ceil|complex|contains|copy|'
                r'deepcopy|del|delattr|delete|delitem|dir|div|divmod|enter|eq|'
                r'exit|float|floor|floordiv|format|ge|get|getattr|getattribute|'
                r'getinitargs|getitem|getnewargs|getstate|gt|hash|hex|iadd|'
                r'iand|idiv|ifloordiv|ilshift|imul|index|init|instancecheck|'
                r'int|invert|ior|ipow|irshift|isub|iter|itruediv|ixor|le|len|'
                r'lshift|lt|metaclass|missing|mod|mul|ne|neg|new|next|oct|or|'
                r'pos|pow|prepare|radd|rand|rdiv|rdivmod|reduce|reduce_ex|'
                r'repr|reversed|rfloordiv|rlshift|rmod|rmul|ror|round|rpow|'
                r'rrshift|rshift|rsub|rtruediv|rxor|set|setattr|setitem|'
                r'setstate|sizeof|str|sub|subclasscheck|subclasshook|truediv|'
                r'trunc|unicode|weakref|xor'
                r')__\b'
            )
        },


#-- STRING --------------------------------------------------------------------#
        # ALL ARE COMMON

#-- REGEX ---------------------------------------------------------------------#
        # ALL ARE COMMON
    })
